### PR TITLE
[GraphQl] In case of scalar types, convert the Iterable into Array

### PR DESCRIPTION
**Pull request by epourail**
…ray of base-type like [StringType] (or [BoolType] or [IntegerType] or [FloatType])

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #2128
| License       | MIT
| Doc PR        | -

This PR sets the graphql type for the fields defined as an array of scalar-type (like `string[]`).
The associated graphql type will be an array of base-type like `[StringType]` (or `[BooleanType]` or `[IntegerType]` or `[FloatType]`) and not an `Iterable`

| PHP type | GraphQl type |
|--|--|
| `@var array` | `Iterable` |
| `@var string[]` | `[StringType]` |
| `@var int[]` | `[IntegerType]` |
| `@var float[]` | `[FloatType]` |
| `@var bool[]` | `[BooleanType]` |

### CODE REVIEW

File: src/GraphQl/Type/SchemaBuilder.php [hunk 1]
```diff
@@ -507,8 +524,30 @@ private function getResourcePaginatedCollectionType(string $resourceClass, Graph
         return $this->graphqlTypes[$resourceClass]['connection'] = new ObjectType($configuration);
     }
 
-    private function isCollection(Type $type): bool
+    private function isCollectionOfObject(Type $type): bool
```
[the remark this answers is not in the document]
Comment (epourail) [reply src/GraphQl/Type/SchemaBuilder.php]: renamed

File: src/GraphQl/Type/SchemaBuilder.php [hunk 2]
```diff
@@ -507,8 +524,30 @@ private function getResourcePaginatedCollectionType(string $resourceClass, Graph
...
+    {
+        return $type->isCollection()
+            && Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType();
+    }
+
+    private function isCollectionOfScalar(Type $type): bool
```
[the remark this answers is not in the document]
Comment (epourail) [reply src/GraphQl/Type/SchemaBuilder.php]: renamed

File: src/GraphQl/Type/SchemaBuilder.php [hunk 3]
```diff
@@ -507,8 +524,30 @@ private function getResourcePaginatedCollectionType(string $resourceClass, Graph
...
+            && null !== $type->getCollectionValueType()
+            && null === $type->getCollectionValueType()->getClassName()
+            && \in_array(
+                $type->getCollectionValueType()->getBuiltinType(),
+                [Type::BUILTIN_TYPE_INT, Type::BUILTIN_TYPE_FLOAT, Type::BUILTIN_TYPE_STRING, Type::BUILTIN_TYPE_BOOL],
+                true
+            );
+    }
+
+    private function convertToIterableType()
```
[the remark this answers is not in the document]
Comment (epourail) [reply src/GraphQl/Type/SchemaBuilder.php]: renamed

**Comment by epourail** [comment]
I want to check more use cases.
The `Dummy` object allows to introspect the `array` will be converted to `Iterable` (via the `arrayData`)
But I want to check too the `string[]` will be converted to `[StringType]

**Comment by alanpoulain** [comment]
My bad, the `arrayData` is indeed a simple array. Did you try to debug the introspection step?

File: src/GraphQl/Type/SchemaBuilder.php [hunk 4]
```diff
@@ -513,8 +530,29 @@ private function getResourcePaginatedCollectionType(string $resourceClass, Graph
...
+    private function isCollectionOfObjects(Type $type): bool
+    {
+        return $type->isCollection()
+            && Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType();
+    }
+
+    private function isCollectionOfScalars(Type $type): bool
     {
-        return $type->isCollection() && Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType();
+        return $type->isCollection()
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: Could you please replace this by:
```
return $type->isCollection()
        return $type->isCollection()
            && null !== $type->getCollectionValueType()
            && null !== $type->getCollectionValueType()
            && null === $type->getCollectionValueType()->getClassName();
```
Unless it doesn't work of course.

Comment (epourail) [reply src/GraphQl/Type/SchemaBuilder.php]: updated.
Note : the unitary test succeeded in local but failed during the "ci/circleci: phpunit-coverage" step due to a slowest test.

File: .php_cs.cache.old [hunk 5]
```diff
@@ -0,0 +1 @@
...":3001950043,"tests\/Hydra\/Serializer\/ConstraintViolationNormalizerTest.php":3773847438,"tests\/Hydra\/Serializer\/PartialCollectionViewNormalizerTest.php":2561475409,"tests\/Security\/EventListener\/DenyAccessListenerTest.php":381632703,"tests\/Security\/ResourceAccessCheckerTest.php":2783221473,"tests\/Identifier\/IdentifierConverterTest.php":3282382818,"tests\/Identifier\/Normalizer\/DateTimeIdentifierNormalizerTest.php":1834994533,"tests\/Identifier\/Normalizer\/IntegerDenormalizerTest.php":4026468164,"tests\/Identifier\/CompositeIdentifierParserTest.php":2551211900,"tests\/Swagger\/Serializer\/DocumentationNormalizerTest.php":4222709532,"tests\/Swagger\/Serializer\/ApiGatewayNormalizerTest.php":505325635,"tests\/Serializer\/SerializerContextBuilderTest.php":2057043402,"tests\/Serializer\/Filter\/PropertyFilterTest.php":110880211,"tests\/Serializer\/Filter\/GroupFilterTest.php":3002831442,"tests\/Serializer\/SerializerFilterContextBuilderTest.php":174625565,"tests\/Serializer\/ItemNormalizerTest.php":860382001,"tests\/Serializer\/JsonEncoderTest.php":3532383839,"tests\/Serializer\/AbstractItemNormalizerTest.php":484713053,"tests\/Serializer\/ResourceListTest.php":2371857971}}
```
Comment (alanpoulain) [inline .php_cs.cache.old]: This file should be removed.

File: src/GraphQl/Type/SchemaBuilder.php [hunk 6]
```diff
@@ -353,23 +355,28 @@ private function convertType(Type $type, bool $input = false, string $mutationNa
...
 
-                $resourceClass = $this->isCollection($type) ? $type->getCollectionValueType()->getClassName() : $type->getClassName();
+                $resourceClass = $this->isCollectionOfObjects($type)
+                    ? $type->getCollectionValueType()->getClassName()
+                    : $type->getClassName();
                 if (null === $resourceClass) {
                     return null;
                 }
-
+
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: Please remove the spaces.

**Comment by alanpoulain** [comment]
Could you try to rebase @epourail?

**Comment by alanpoulain** [comment]
After some investigation, I think this PR is not useful for now.
It's because all the array types of the `DummyIterable` have `string` for the collection value type.
Why is that? It's because property metadata are generated using the `PropertyInfoPropertyMetadataFactory`. The `PropertyInfoExtractor` of Symfony has this method:
https://github.com/symfony/symfony/blob/af9920cf1078643a8f36df0fd4c2e8704370ce04/src/Symfony/Component/PropertyInfo/PropertyInfoExtractor.php#L101-L113
As you can see, it's the first extractor returning a result that is used. In our case, it's the `DoctrineExtractor`.
And here we have the issue:
https://github.com/symfony/symfony/blob/af9920cf1078643a8f36df0fd4c2e8704370ce04/src/Symfony/Bridge/Doctrine/PropertyInfo/DoctrineExtractor.php#L166
So if the field is a `simple_array` the `collectionValueType` will always be the same, whatever you put in the PhpDoc block.
IMO, it's a hard issue to solve in Symfony (cc @dunglas).

**Comment by flug** [comment]
@alanpoulain if the pull request is invalid maybe it's time to close it? @epourail

**Comment by soyuka** [comment]
Closing feel free to discuss this further

**Pull request closed without merging**